src/core/namespace.c

```diff
@@ -3181,6 +3183,117 @@ bool ns_type_supported(NamespaceType type) {
 }
 
+int refresh_extensions_in_namespace(
+                const PidRef *target,
+                const char *hierarchy_env,
+                const NamespaceParameters *p) {
+
+        const char *overlay_prefix = "/run/systemd/mount-rootfs";
+        _cleanup_(mount_list_done) MountList ml = {};
+        _cleanup_free_ char *extension_dir = NULL;
+        _cleanup_strv_free_ char **hierarchies = NULL;
+        MountInNamespaceFlags min_flags = 0;
+        int r;
+
+        assert(pidref_is_set(target));
+        assert(hierarchy_env);
+        assert(p);
+
+        log_debug("Refreshing extensions in-namespace for hierarchy '%s'", hierarchy_env);
+
+        extension_dir = path_join(p->private_namespace_dir, "unit-extensions");
+        if (!extension_dir)
+                return -ENOMEM;
+
+        min_flags |= MOUNT_IN_NAMESPACE_MAKE_FILE_OR_DIRECTORY;
+
+        r = parse_env_extension_hierarchies(&hierarchies, hierarchy_env);
+        if (r < 0)
+                return r;
+
+        r = append_extensions(
+                        &ml,
+                        overlay_prefix,
+                        p->private_namespace_dir,
+                        hierarchies,
+                        p->extension_images,
+                        p->n_extension_images,
+                        p->extension_directories);
+        if (r < 0)
+                return r;
+        if (ml.n_mounts == 0)
+                return 0;
+
+        r = safe_fork("(sd-ns-refresh-exts)",
+                        FORK_DEATHSIG_SIGTERM | FORK_WAIT | FORK_NEW_MOUNTNS | FORK_MOUNTNS_SLAVE,
+                        NULL);
+        if (r < 0)
+                return r;
+        if (r == 0) {
+                (void) mkdir_p_label(overlay_prefix, 0555);
+
+                /* This is effectively two rounds, since all the extensions come before overlays
+                * (setup_namespace() similarly relies on this property).
+                *
+                * (1) First, set up all the extension mounts in the child, which are not visible from the
+                * process. (2) Then, set up overlays for the sysext/confext hierarchies again using the new
+                * extension mounts as layers, and move them into the namespace. */
+                FOREACH_ARRAY(m, ml.mounts, ml.n_mounts) {
+                        if (IN_SET(m->mode, MOUNT_EXTENSION_DIRECTORY, MOUNT_EXTENSION_IMAGE)) {
+                                r = apply_one_mount(p->root_directory, m, p);
+                                if (r < 0) {
+                                        log_debug_errno(r, "Failed to apply extension mount: %m");
+                                        _exit(EXIT_FAILURE);
+                                }
+                        } else if (m->mode == MOUNT_OVERLAY) {
+                                _cleanup_free_ char *path_relative = NULL, *path_in_namespace = NULL;
+
+                                r = apply_one_mount(p->root_directory, m, p);
```

Comment:
so, hmm, i don't follow here, we need to include the "real" /etc/ in the layers here, but where do we get that from here? we need to pull it out of the existing service namespace, but i don't see that hapenning here? what am i missing?

Reply:
I think root_directory here is misleading and should be dropped. What happens is that when the MOUNT_OVERLAY entry is applied, that namespace's /etc/ is added as the last element in lowerdir=. But I think we are applying this in the wrong namespace, the root one, so we end up using the host's /etc/?

So I guess what we need to do is set up the extensions in the new mount namespace derived from the host's as we do now, so that we see them, but apply the overlay in the final one so that its /etc/ is picked up as the base? But would we have permissions to do so?

Reply:
at the moment you apply the overlayfs you need access to all extension dirs and of the base /etc, so much is clear.

So after discussing briefly with Christian this apparently can be done via /proc/self/fd/<nr> references in lowerdirs.

i.e. the code should open all extension roots on the host, then enter the target namespace, also open the underlying /etc/ from there, and then combine extensions + underlying /etc/ in one overlayfs via lowerdirs= all referencing /proc/self/fd/ paths.

that would address relatively nicely that we need some resources from the host and some from the service namespace.

Reply:
If we are in the target namespace already, is there any reason to take a fd reference of the rootfs instead of just using the path? It would be the same, no?

But also, do we actually have permission to create a new overlayfs if we are in the target namespace? I know we need to be in the host's to mount the images, but I am unsure about the overlay

Reply:
If we are in the target namespace already, is there any reason to take a fd reference of the rootfs instead of just using the path? It would be the same, no?

well, you need to get access to the rootfs first, which is a bit hairy. You need to create a throw-away namespace for that, then "unpeal" /etc (i.e. get rid of the old overlayfs on it by unmounting), then getting an fd on that dir, and send it back to your worker process.

Reply:
> But also, do we actually have permission to create a new overlayfs if we are in the target namespace? I know we need to be in the host's to mount the images, but I am unsure about the overlay

mount namespaces carry no privileges. if you are root outside you are root inside. 

(userns would be different, but don#t bother with userns)

Reply:
> well, you need to get access to the rootfs first, which is a bit hairy. You need to create a throw-away namespace for that, then "unpeal" /etc (i.e. get rid of the old overlayfs on it by unmounting), then getting an fd on that dir, and send it back to your worker process.

But in this case we don't need access to the rootfs beforehand, as we want to create the overlay directly in the target namespace, no? So doing `lowerdir=...:/etc/` should be enough, as that refers to the right thing already, right?

Reply:
hmm? /etc/ would refer to the previous overlayfs. so you'd create an overlayfs whose one layer would be another overlayfs. Do that 7x and you 7 overlayfs in your stack

(that said, I think overlayfs refuses such recursive uses, so you wouldn't even get that far I think)

Reply:
But doesn't that apply to proc/self/fd just as well? How do you get a fd to the 'original' /etc/ without separately remounting the rootfs?

Reply:
as i wrote above. here's the combined logic:

0. you open all extensions images/directories in the parent (either via new mount api stuff, or via shortlived mount namespaces for the DDI cases)
1. you create a child process that now has fds to the extension images/dirs and that now enters the service's mount ns
2. you then create a grandchild process, clone the mount namespace of the child in it, and disable propagation in it
3. you unmount /etc/ as long as it is a confext overlayfs in the grandchild
4. you open the now "unpealed" /etc/ in the grandchild, and send the fd of it to the child
5. you exit in the grandchild
6. the child now has all fds it needs, it now mounts them to /etc/ as a new overlayfs based on only fds, using the beneath logic to make things atomic
7. the child exits too now
8. …
9. profitt!

Reply:
you should really reuse the existing superblock of the base /etc/, i.e. get it out of the existing namespace.

if you set things up anew you might end up with completely different superblocks, and that sounds like a problem waiting to happen.

Ideally overlayfs would provide us with some ioctl()s so that we could get access to it's layers directly, but it doesn't. If it did you could just query the base layer of it (after all the lowest layer in the overlayfs is identical to the dir we mount over).

Reply:
we need to make a copy of the mount though, no? otherwise the umount loop will affect the live process?